qa/L0_response_cache/ensemble_cache_test.py

```diff
@@ -0,0 +1,182 @@
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import os
+import re
+import sys
+
+sys.path.append("../common")
+sys.path.append("../clients")
+
+import logging
+import unittest
+
+import infer_util as iu
+import numpy as np
+import test_util as tu
+import tritonclient.grpc as grpcclient
+
+
+class EnsembleCacheTest(tu.TestResultCollector):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.ensemble_model_name = "simple_graphdef_float32_float32_float32"
+        self.composing_model_name = "graphdef_float32_float32_float32"
+        self.model_directory = os.path.join(os.getcwd(), "models", "ensemble_models")
+        input0_data = np.ones((1, 16), dtype=np.float32)
+        input1_data = np.ones((1, 16), dtype=np.float32)
+        self.input_tensors = [
+            grpcclient.InferInput("INPUT0", [1, 16], "FP32"),
+            grpcclient.InferInput("INPUT1", [1, 16], "FP32"),
+        ]
+        self.input_tensors[0].set_data_from_numpy(input0_data)
+        self.input_tensors[1].set_data_from_numpy(input1_data)
+
+    def _get_infer_result(self):
+        triton_client = grpcclient.InferenceServerClient("localhost:8001", verbose=True)
+        results = triton_client.infer(
+            model_name=self.ensemble_model_name, inputs=self.input_tensors
+        )
+        stats = triton_client.get_inference_statistics(
+            model_name=self.ensemble_model_name, as_json=True
+        )
+        return stats
+
+    def _run_ensemble(self):
+        stats = self._get_infer_result()
```

Comment:
rather than doing two inferences in this function , discarding the first stats, and checking the second stats in each test - we should probably do something closer to this for each test:
1. check stats are all 0
2. do inference
3. check cache miss is 1, and cache hit is 0
4. do inference
5. check cache miss is 1, check cache hit is 1

What do you think?

Reply:
Ideally, we should do this. An issue that I encountered was with trying to kill Triton server which. When I add the lines
kill $SERVER_ID
wait $SERVER_ID
The wait function is not returning or maybe returning with an error code, because after wait command, the other lines are not executing.
Without killing the server and re-launching it, the stats won't be empty.
Let me know if my understanding is wrong.

Reply:
Are these comments meant for the other thread about removing the `wait $SERVER_PID`? I don't see the total relevance of the `wait` function for this thread. If you wanted to avoid starting/stopping server between tests there are other things you can do too. Rather than assert they are all zero, you can assert they are all some expected baseline value (based on test knowledge or something - but this may be fragile if it depends on test order). Or you can load the models under new names, or just unload/load the same model, etc. - you can do things to clear the stats if it helps improve the test sanity.

Reply:
Ok, will fix it.